server/src/com/mirth/connect/server/userutil/VMRouter.java

```diff
@@ -119,11 +167,177 @@ public Response routeMessageByChannelId(String channelId, RawMessage rawMessage)
+        return routeMessageByChannelId(channelId, createRawMessage(message, sourceMap, destinationSet));
+    }
+
     private com.mirth.connect.donkey.model.message.RawMessage convertRawMessage(RawMessage message) {
         if (message.isBinary()) {
             return new com.mirth.connect.donkey.model.message.RawMessage(message.getRawBytes(), message.getDestinationMetaDataIds(), message.getSourceMap());
         } else {
             return new com.mirth.connect.donkey.model.message.RawMessage(message.getRawData(), message.getDestinationMetaDataIds(), message.getSourceMap());
         }
     }
+
+    /**
+        Create a {@link RawMessage} with the specified content, sourceMap, and destinationSet.
+
+        @param message The content of the message to be sent, textual or binary. As String or byte[].
+        @param sourceMap A map containing entries to include in the sourceMap of the {@link RawMessage} (optional).
+        @param destinationSet A collection of integers (metadata IDs) representing which destinations to dispatch the message to.
+            Null may be passed to indicate all destinations. If unspecified, all destinations is the default (optional).
+        @return A {@link RawMessage} object containing the message, source, and destination information.
+    */
+    public RawMessage createRawMessage(Object message, Map<String, Object> sourceMap, Collection<Number> destinationSet) {
+        if(trackingEnhancer != null) {
+            sourceMap = trackingEnhancer.enrich(sourceMap);
+        }
+
+        if(message instanceof byte[]) {
+            return new RawMessage((byte[])message, destinationSet, sourceMap);
+        } else {
+            return new RawMessage(message.toString(), destinationSet, sourceMap);
+        }
+    }
+
+    /**
+     * Adds additional message tracking data.
+     * 
+     * TrackingEnhancer
+     */
+    private class TrackingEnhancer {
+        private String channelId;
+        private Long messageId;
+        private SourceMap envSourceMap;
+        
+        /**
+         * Constructor
+         * 
+         * @param channelId channel ID; null defaults to "NONE"
+         * @param messageId message ID; null defaults to -1L
+         * @param sourceMap the message's source map
+         */
+        private TrackingEnhancer(String channelId, Long messageId, SourceMap sourceMap) {
+            this.channelId = channelId != null ? channelId : "NONE";
+            this.messageId = messageId != null ? messageId : -1L;
+            this.envSourceMap = sourceMap;
+        }
+
+        /**
+         * Enrich the given source map with additional message tracking data.
+         * 
+         * @param messageSourceMap
+         * @return a new Map
+         */
+        private Map<String, Object> enrich(Map<String, Object> messageSourceMap) {
+            if (messageSourceMap == null) {

```

Comment:
Yes, I see it can come in as null from at least one method.

I don't see other places where `@Null` and `@NotNull` are used that I could copy, so I defer to others on properly decorating and validating the code.